improv(recipe): Added blockstate recipe mcfunction support

00-Steven opened the pull request:
Adds mcfunction support to the existing BlockState recipe system.

---

I agree to the Contributor License Agreement (CLA).

00-Steven committed:
improv(recipe): Added blockstate recipe mcfunction support

Added mcfunction support to most of the blockstate recipes.

bconlon1 commented on src/main/java/com/gildedgames/aether/data/providers/AetherRecipeProvider.java:
```diff
@@ -416,6 +417,10 @@ public BlockStateRecipeBuilder ambrosiumEnchanting(Block result, Block ingredien
         return BlockStateRecipeBuilder.recipe(BlockStateIngredient.of(ingredient), result, AetherRecipeSerializers.AMBROSIUM_ENCHANTING.get());
     }
 
+    public BlockStateRecipeBuilder ambrosiumEnchanting(Block result, Block ingredient, ResourceLocation mcfunction) {
```
This is probably not necessary considering it'd be unused by our data generators.

bconlon1 commented on src/main/java/com/gildedgames/aether/recipe/builder/BiomeParameterRecipeBuilder.java:
```diff
@@ -24,38 +24,62 @@ public class BiomeParameterRecipeBuilder extends BlockStateRecipeBuilder {
 
-    public static BiomeParameterRecipeBuilder recipe(BlockPropertyPair result, BlockStateIngredient ingredient, @Nullable ResourceKey<Biome> biomeKey, @Nullable TagKey<Biome> biomeTag, BlockStateRecipeSerializer<?> serializer) {
-        return new BiomeParameterRecipeBuilder(result, ingredient, biomeKey, biomeTag, serializer);
+    public static BiomeParameterRecipeBuilder recipe(BlockStateIngredient ingredient, Block result, ResourceKey<Biome> biomeKey, BlockStateRecipeSerializer<?> serializer, ResourceLocation mcfunction) {
```
The amount of methods could probably be cut down by having a new method to handle appending an mcfunction through the builder, similar to how vanilla does for things like groups:
```
public ShapedRecipeBuilder mcFunction(@Nullable ResourceLocation mcFunction) {
  this.mcFunction = mcFunction;
  return this;
}
```

bconlon1 replied:
Or well, doing that could probably be done in BlockStateRecipeBuilder instead which would cut down on even more changes.

bconlon1 commented on src/main/java/com/gildedgames/aether/recipe/serializer/BiomeParameterRecipeSerializer.java:
```diff
@@ -40,7 +41,7 @@ public T fromNetwork(@Nonnull ResourceLocation recipeId, @Nonnull FriendlyByteBu
         BlockStateIngredient ingredient = BlockStateIngredient.fromNetwork(buf);
         BlockPropertyPair result = BlockStateRecipeUtil.readPair(buf);
-        return this.factory.create(recipeId, biomeKey, biomeTag, ingredient, result);
+        return this.factory.create(recipeId, biomeKey, biomeTag, ingredient, result, CommandFunction.CacheableFunction.NONE);
```
I believe you need to handle the networking for the function instead of it being empty or it may not work on multiplayer, but I'd need a datapack to test with to be certain.

bconlon1 reviewed (changes requested):
Most of this PR looks good aside from the attached requested changes. I have nitpicks like the capitalization of `mcfunction` or `Mcfunction` (but honestly I'm not sure what I'd do with it), and the usage of `var`, and it might be possible to abstract some of the code segments so they can be used through method calls instead, but I might just handle that myself since those aren't really implementation issues.

Burning-Cactus committed:
feat(structure): Add the bronze dungeon to the Aether (#811)

Implements the Bronze Dungeon as close to completion as possible; beardifier usage is temporary and dungeon doesn't properly overwrite features inside it yet

closes issue #225

reviewed by bconlon

merged from #811

Burning-Cactus committed:
improv(entity): Valkyrie and Valkyrie Queen improvements (#854)

Valkyries now have their lunge attack back
Valkyries no longer get stuck in boats
Valkyrie queens will teleport up to the player if the player is too high for them to reach for too long
Partially addresses #776

reviewed by bconlon

merged from #854

Burning-Cactus committed:
fix(entity): Rework valkyrie lunge and increase slider dungeon bounds (#855)

Valkyries now take knockback into account instead of juggernauting through.
Bronze dungeons properly set the slider's dungeon bounds now.

reviewed by bconlon

merged from #855

bconlon1 committed:
chore: Edit CONTRIBUTING.md

[ci skip]

bconlon1 committed:
feat(item): More item fixes and additions (#858)

Add optional tags allowing other Curios types to be put in the Aether's accessory menu, add cooldowns for the Cloud Staff and Hammer of Notch usage, fixed Valkyrie Armor's wings having buggy rotation that played even when paused, and fixed Capes having buggy rotation from not syncing in multiplayer

reviewed by Raptor

merged from #858

bconlon1 committed:
feat(block)!: Additions, fixes, and general cleanup of block code (#856)

Cleaned up block package, blockentity package, portal code, Aether, AetherConfig, AetherTags, access transformers by switching to mixin accessors, and various mixins

BREAKING CHANGE: Various classes have been renamed and moved around

reviewed by Burning-Cactus

merged from #856

bconlon1 committed:
build(version): 0.1.0 -> 0.2.0

bconlon1 committed:
refactor: Balanced icestone fuel values (#859)

Sets the Icestone Stairs, Slab, and Wall to the right fuel values in the
Freezer

reetamb committed:
feat(structure): Dungeon Blacklist Placement Filter (#863)

Adds a PlacementFilter that prevents Features from spawning within the bounds of a structure. This is applied onto Crystal Trees, Base Trees, Springs, Quicksoil Shelves, and Aerclouds

closes #790

reviewed by bconlon

merged from #863

bconlon1 committed:
chore: Remove README email references

bconlon1 committed:
chore: Remove AGREEMENT email references

[ci skip]

Commit:
docs: Update FreezingBehavior docs

Fixes merge conflict

Commit:
feat: Code cleanup

Commit:
style: Remove space

bconlon1 reviewed (approved):
lgtm

Aether-Team-Bot committed:
chore: Trigger build

Aether-Team-Bot committed:
Merge branch 'local/00-Steven/feat/00-Steven/mcfunctionsupport' into merge/feat/00-Steven/mcfunctionsupport
